### ISSUE TITLE
Presentation of decorators

### ISSUE THREAD

**Issue by mthuurne** [twisted/pydoctor#318]
This is a collection of ideas on how we can improve the presentation of decorators in the generated documentation.

### Private vs public

Some decorators express something about the decorated object that is part of the API, while others are implementation helpers that should be hidden under "Show Private API".

One obvious thing we could look at is the decorator function itself: if it's a private function by naming convention or being defined in a private module, we can safely assume it's an implementation helper.

But there could also be cases where implementation helpers are public because the decorator is useful to end users or because it comes from an external library, but the decoration (the `@` expression) should not be considered part of the public interface of the project. In that case it would be useful if we had a way to mark a decorator as private. For internal decorators this could be done via a docstring field, but for external decorators we'd need a different mechanism.

Maybe there are also cases where we want the decoration itself to be visible, but want to hide the arguments passed to it.

### Names in decorations

Currently decorations are produced by feeding the decoration's AST (the expression after `@`) to `astor` and displaying the resulting text. It would be more useful to format them in a way similar to how type annotations are formatted, where names of functions etc. are linked to their definitions (if available internally or via intersphinx).

Related to this, it would be useful to normalize the presentation of the name itself. For example, in Twisted some modules use `@deprecated` and others use `@deprecate.deprecated`, but both decorations are doing the same thing and it would be better to format them in the same way in the documentation (if this decoration wasn't suppressed; it's just an example). I posed the question of how to present type names in #280, but I guess it applies to other kinds of names as well.

### List decorated objects

In the documentation of the decorator itself, it would be useful to include a list of decorated objects, similar to how we keep track of which classes implement each interface.

**Comment by tristanlatr** [comment]
> if it's a private function by naming convention or being defined in a private module, we can safely assume it's an implementation helper.

I agree. Now, I wouldn't reinvent the wheel in terms of private/public handling. No idea how we can "mark a decorator as private", and I'm unsure if that functionality would ben actually used. I think we should keep this simple. 

> Maybe there are also cases where we want the decoration itself to be visible, but want to hide the arguments passed to it.

In my opinoin, this also goes also in the "that's too much to handle" zone. 

> It would be more useful to format them in a way similar to how type annotations are formatted, where names of functions etc. are linked to their definitions 

Once #402 is merged, this task will be pretty easy with the new PyvalColorizer. 

> In the documentation of the decorator itself, it would be useful to include a list of decorated objects, similar to how we keep track of which classes implement each interface.

That would be nice indeed.

**Comment by tristanlatr** [comment]
I just realized that decorator on classes are not shown, so showing them in the same fashion we show them for functions would be an improvement.

**Comment by tristanlatr** [comment]
Follow-up on this issue: The current behaviour of pydoctor with function decorators is to syntax highlight them with the colorizer whether there are private or public. 

And we're not tracking which object is decorated with which decorator. Which would actually not be very hard, but I wonder what would be the real life benefits.
